webhookValidator_test.go

```diff
@@ -0,0 +1,237 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ancla
+
+import (
+	"errors"
+	"net/url"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+var (
+	_ Validator = GoodURL([]ValidURLFunc{HTTPSOnlyEndpoints(), RejectHosts(nil), RejectAllIPs()})
+)
+
+func TestGoodURL(t *testing.T) {
+	tcs := []struct {
+		Desc        string
+		URL         Webhook
+		ExpectedErr error
+	}{
+		{
+			Desc: "Blank String",
+			URL: Webhook{Config: DeliveryConfig{URL: ""},
+				FailureURL: ""},
+			ExpectedErr: errInvalidURL,
+		},
+		{
+			Desc: "No https URL",
+			URL: Webhook{Config: DeliveryConfig{URL: "http://www.google.com/"},
+				FailureURL: "https://www.google.com/"},
+			ExpectedErr: errInvalidURL,
+		},
+		{
+			Desc: "Good case",
+			URL: Webhook{Config: DeliveryConfig{URL: "https://www.google.com/"},
+				FailureURL: "https://www.google.com:1030/software/index.html"},
+		},
+		{
+			Desc: "Bad Failure URL scheme",
+			URL: Webhook{Config: DeliveryConfig{URL: "https://www.google.com/"},
+				FailureURL: "127.0.0.1:1030"},
+			ExpectedErr: errInvalidFailureURL,
+		},
+		{
+			Desc: "Bad Failure URL",
+			URL: Webhook{Config: DeliveryConfig{URL: "https://www.google.com/"},
+				FailureURL: "https://127.0.0.1:1030"},
+			ExpectedErr: errInvalidFailureURL,
+		},
+		{
+			Desc: "Good case",
+			URL: Webhook{
+				Config: DeliveryConfig{
+					URL:             "https://www.google.com/",
+					AlternativeURLs: []string{"https://www.google.com/", "https://www.bing.com/"}},
+				FailureURL: "https://www.google.com:1030/software/index.html"},
+		},
+		{
+			Desc: "Bad Alternative URLs",
+			URL: Webhook{
+				Config: DeliveryConfig{
+					URL:             "https://www.google.com/",
+					AlternativeURLs: []string{"https://www.google.com/", "http://www.bing.com/"}},
+				FailureURL: "https://www.google.com:1030/software/index.html"},
+			ExpectedErr: errInvalidAlternativeURL,
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.Desc, func(t *testing.T) {
+			assert := assert.New(t)
+			err := GoodURL([]ValidURLFunc{HTTPSOnlyEndpoints(), RejectAllIPs()})(tc.URL)
+			assert.True(errors.Is(err, tc.ExpectedErr))
+		})
+	}
+}
+
+func TestHTTPSOnlyEndpoints(t *testing.T) {
+	tcs := []struct {
+		Desc        string
+		URL         string
+		ExpectedErr error
+	}{
+		{
+			Desc:        "No https URL",
+			URL:         "http://www.google.com/",
+			ExpectedErr: errURLIsNotHTTPS,
+		},
+		{
+			Desc:        "Invalid host with Port",
+			URL:         "http://www.example.com:1030/software/index.html",
+			ExpectedErr: errURLIsNotHTTPS,
+		},
+		{
+			Desc: "Loopback URL with Port",
+			URL:  "https://localhost:9000",
+		},
```

Comment:
Can you add another test where we pass something that is parsed but doesn't provide a scheme? ie `/example/test`